File: src/audio/yin.ts
/**
 * YIN pitch detector (de Cheveigné & Kawahara, 2002).
 * Returns the fundamental frequency and a 0–1 confidence score.
 */
export interface YinResult {
  frequency: number | null;
  probability: number;
  tau: number | null;
}

export interface YinOptions {
  threshold?: number;
  minFreq?: number;
  maxFreq?: number;
}

function parabolicInterpolation(buffer: Float32Array, tau: number): number {
  const x0 = tau < 1 ? tau : tau - 1;
  const x2 = tau + 1 < buffer.length ? tau + 1 : tau;
  if (x0 === tau) return tau;
  if (x2 === tau) return tau;
  const s0 = buffer[x0];
  const s1 = buffer[tau];
  const s2 = buffer[x2];
  const denom = 2 * s1 - s2 - s0;
  if (denom === 0) return tau;
  return tau + (s2 - s0) / (2 * denom);
}

export function detectPitchYin(
  samples: Float32Array,
  sampleRate: number,
  options: YinOptions = {},
): YinResult {
  const threshold = options.threshold ?? 0.12;
  const minFreq = options.minFreq ?? 55; // A1 — beginners rarely go lower
  const maxFreq = options.maxFreq ?? 2093; // C7
  const half = Math.floor(samples.length / 2);
  const tauMin = Math.max(2, Math.floor(sampleRate / maxFreq));
  const tauMax = Math.min(half - 1, Math.floor(sampleRate / minFreq));
  if (tauMax <= tauMin + 2) {
    return { frequency: null, probability: 0, tau: null };
  }

  const yin = new Float32Array(tauMax + 1);
  for (let tau = 1; tau <= tauMax; tau++) {
    let sum = 0;
    for (let i = 0; i < half; i++) {
      const delta = samples[i] - samples[i + tau];
      sum += delta * delta;
    }
    yin[tau] = sum;
  }

  yin[0] = 1;
  let runningSum = 0;
  for (let tau = 1; tau <= tauMax; tau++) {
    runningSum += yin[tau];
    yin[tau] = runningSum === 0 ? 1 : (yin[tau] * tau) / runningSum;
  }

  let tauEstimate = -1;
  for (let tau = tauMin; tau <= tauMax; tau++) {
    if (yin[tau] < threshold) {
      while (tau + 1 <= tauMax && yin[tau + 1] < yin[tau]) {
        tau += 1;
      }
      tauEstimate = tau;
      break;
    }
  }

  if (tauEstimate === -1) {
    let best = 1;
    let bestTau = -1;
    for (let tau = tauMin; tau <= tauMax; tau++) {
      if (yin[tau] < best) {
        best = yin[tau];
        bestTau = tau;
      }
    }
    if (bestTau === -1 || best > 0.35) {
      return { frequency: null, probability: 0, tau: null };
    }
    tauEstimate = bestTau;
  }

  const betterTau = parabolicInterpolation(yin, tauEstimate);
  if (betterTau <= 0) {
    return { frequency: null, probability: 0, tau: null };
  }
  const probability = Math.max(0, Math.min(1, 1 - yin[tauEstimate]));
  return {
    frequency: sampleRate / betterTau,
    probability,
    tau: betterTau,
  };
}

export function rms(samples: Float32Array): number {
  let sum = 0;
  for (let i = 0; i < samples.length; i++) {
    sum += samples[i] * samples[i];
  }
  return Math.sqrt(sum / samples.length);
}

File: android/core/pitch/src/main/kotlin/com/vijaychhetry/kidspiano/core/pitch/Yin.kt
package com.vijaychhetry.kidspiano.core.pitch

import com.vijaychhetry.kidspiano.core.common.AudioFrame
import com.vijaychhetry.kidspiano.core.common.PitchResult

interface PitchDetector {
    fun detect(frame: AudioFrame): PitchResult
}

data class YinEstimate(val frequency: Double?, val probability: Double)

fun rms(samples: FloatArray): Double {
    var sum = 0.0
    for (s in samples) sum += s * s
    return Math.sqrt(sum / samples.size)
}

fun detectPitchYin(
    samples: FloatArray,
    sampleRate: Int,
    threshold: Double = 0.12,
    minFreq: Double = 130.0,
    maxFreq: Double = 800.0,
): YinEstimate {
    val half = samples.size / 2
    val tauMin = maxOf(2, Math.floor(sampleRate / maxFreq).toInt())
    val tauMax = minOf(half - 1, Math.floor(sampleRate / minFreq).toInt())
    if (tauMax <= tauMin + 2) return YinEstimate(null, 0.0)

    val yin = DoubleArray(tauMax + 1)
    for (tau in 1..tauMax) {
        var sum = 0.0
        for (i in 0 until half) {
            val delta = samples[i] - samples[i + tau]
            sum += delta * delta
        }
        yin[tau] = sum
    }
    yin[0] = 1.0
    var running = 0.0
    for (tau in 1..tauMax) {
        running += yin[tau]
        yin[tau] = if (running == 0.0) 1.0 else (yin[tau] * tau) / running
    }

    var tauEstimate = -1
    var tau = tauMin
    while (tau <= tauMax) {
        if (yin[tau] < threshold) {
            while (tau + 1 <= tauMax && yin[tau + 1] < yin[tau]) tau++
            tauEstimate = tau
            break
        }
        tau++
    }
    if (tauEstimate == -1) {
        var best = 1.0
        var bestTau = -1
        for (t in tauMin..tauMax) {
            if (yin[t] < best) {
                best = yin[t]
                bestTau = t
            }
        }
        if (bestTau == -1 || best > 0.35) return YinEstimate(null, 0.0)
        tauEstimate = bestTau
    }

    val betterTau = parabolic(yin, tauEstimate)
    if (betterTau <= 0) return YinEstimate(null, 0.0)
    return YinEstimate(
        frequency = sampleRate / betterTau,
        probability = (1.0 - yin[tauEstimate]).coerceIn(0.0, 1.0),
    )
}

private fun parabolic(buffer: DoubleArray, tau: Int): Double {
    val x0 = if (tau < 1) tau else tau - 1
    val x2 = if (tau + 1 < buffer.size) tau + 1 else tau
    if (x0 == tau || x2 == tau) return tau.toDouble()
    val s0 = buffer[x0]
    val s1 = buffer[tau]
    val s2 = buffer[x2]
    val denom = 2 * s1 - s2 - s0
    if (denom == 0.0) return tau.toDouble()
    return tau + (s2 - s0) / (2 * denom)
}
